### PR TITLE
Initial implementation of import-inputs command

### PR DESCRIPTION
**Pull request by tpdownes**
Initial implementation of import-inputs

- read metadata to identify intergroup outputs necessary for target
  deployment group
- collect output values from prior groups and combine into a single file
  in the target deployment group directory

Does not yet support the evaluation of module settings that is necessary for writing to Packer deployment groups. This is anticipated in a follow-up PR now that #1205 has been merged.

Suggest testing with `igc_tf_test.yaml` under tools directory. With this approach one would run:

```shell
terraform -chdir=igc-tf-test/zero init
terraform -chdir=igc-tf-test/zero validate
terraform -chdir=igc-tf-test/zero apply
ghpc export-outputs igc-tf-test/zero

ghpc import-inputs igc-tf-test/one
terraform -chdir=igc-tf-test/one init
terraform -chdir=igc-tf-test/one validate
terraform -chdir=igc-tf-test/one apply
```

Edge cases to consider (among others):

- forget to run export-outputs (or any of the commands for first group) before import-inputs on second group
- attempt to run export-outputs without having run apply
- delete deployment metadata

None of these failures are able to be automatically recovered but the user shouldn't be left helpless.

### CODE REVIEW

File: pkg/shell/terraform.go [hunk 1]
```diff
@@ -176,6 +181,38 @@ func ExportOutputs(tf *tfexec.Terraform, metadataFile string, artifactsDir strin
...
+
+	allAttributes := make(map[string]cty.Value)
+	for group, outputs := range outputsByGroup {
+		if len(outputs) == 0 {
+			continue
+		}
+		filepath := outputsFile(artifactsDir, group)
+		attrs, err := modulereader.ReadHclAttributes(filepath)
+		if err != nil {
+			return fmt.Errorf("could not load file %s; consider running \"ghpc export-outputs %s/%s\".\n%v", filepath, deploymentRoot, group, err)
```
Comment (tpdownes) [inline pkg/shell/terraform.go:205]: @nick-stroud I realize this is probably a good candidate for `TfError`. Will update after review.

**Review by nick-stroud — CHANGES_REQUESTED** [review]

File: pkg/shell/terraform.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/shell/terraform.go:205]: The replacement on HEAD will read:

```shell
2023/05/01 10:39:17 collecting outputs for group one from group zero
Error: consider running "ghpc export-outputs igc-tf-test/zero" (detailed error below)
The configuration file "igc-tf-test/.ghpc/zero_outputs.tfvars" could not be read.
```

File: pkg/shell/terraform.go [hunk 2]
```diff
@@ -176,6 +181,38 @@ func ExportOutputs(tf *tfexec.Terraform, metadataFile string, artifactsDir strin
...
+	outputsByGroup, err := getOutputsFromEarlierGroups(thisGroup, metadataFile)
+	if err != nil {
+		return err
+	}
+
+	if _, err = GetDeploymentKinds(metadataFile, deploymentRoot); err != nil {
+		return err
+	}
+
+	allAttributes := make(map[string]cty.Value)
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/shell/terraform.go]: Similar suggestions implemented.

File: pkg/shell/common.go [hunk 3]
```diff
@@ -64,6 +66,48 @@ func loadMetadata(metadataFile string) ([]modulewriter.GroupMetadata, error) {
 	return md.DeploymentMetadata, nil
 }
 
+// return a map from group names to a list of outputs that are needed by this group
+func getOutputsFromEarlierGroups(thisGroup string, metadataFile string) (map[string][]string, error) {
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/shell/common.go]: Replaced with `getIntergroupOutputNamesByGroup`

File: pkg/shell/terraform.go [hunk 4]
```diff
@@ -176,6 +181,38 @@ func ExportOutputs(tf *tfexec.Terraform, metadataFile string, artifactsDir strin
...
+func ImportInputs(workingDir, metadataFile string, artifactsDir string) error {
+	deploymentRoot := path.Clean(path.Join(workingDir, ".."))
+	thisGroup := path.Base(workingDir)
+
+	outputsByGroup, err := getOutputsFromEarlierGroups(thisGroup, metadataFile)
+	if err != nil {
+		return err
+	}
+
+	if _, err = GetDeploymentKinds(metadataFile, deploymentRoot); err != nil {
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/shell/terraform.go:196]: - function comment for `GetDeploymentKinds` improved
- anticipated import noted as TODO next to where it is called

File: pkg/shell/terraform.go [hunk 5]
```diff
@@ -176,6 +181,38 @@ func ExportOutputs(tf *tfexec.Terraform, metadataFile string, artifactsDir strin
 // ImportInputs will search artifactsDir for files produced by ExportOutputs and
 // combine/filter them for the input values needed by the group in the Terraform
 // working directory
-func ImportInputs(tf *tfexec.Terraform, metadataFile string, artifactsDir string) error {
+func ImportInputs(workingDir, metadataFile string, artifactsDir string) error {
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/shell/terraform.go]: Done

[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/shell/terraform.go]: Resolved by adding type to all there. An edge case to how `golint` enforces function signatures.

**Review by nick-stroud — APPROVED** [review]
